notebooks/09_Monitoring/src/07_01_Iris_Pipeline_Monitoring/model.py

```diff
@@ -0,0 +1,216 @@
+    model.fit(X_train, y_train)
+
+    return model
+
+
+def inference(model, X_test, y_test):
+    y_pred = model.predict(X_test)
+    logging.info("Trained Model's evaluation score: {}".format(model.score(X_test, y_test)))
+    df = pd.concat([pd.DataFrame(y_test, columns=['target']), pd.DataFrame(y_pred, columns=['predicted'])], axis=1)
+    vocab = list(df['target'].unique())
+    cm = confusion_matrix(df['target'], df['predicted'], labels=vocab)
+    data = []
+    for target_index, target_row in enumerate(cm):
+        for predicted_index, count in enumerate(target_row):
+            data.append((vocab[target_index], vocab[predicted_index], count))
+    df_cm = pd.DataFrame(data, columns=['target', 'predicted', 'count'])
+    cm_file = os.path.join('/', 'confusion_matrix.csv')
+    with file_io.FileIO(cm_file, 'w') as f:
+        df_cm.to_csv(f, columns=['target', 'predicted', 'count'], header=False, index=False)
+
+    y_score = model.decision_function(X_test)
+    # Compute ROC curve and ROC area for each class
+    fpr, tpr, thresholds = roc_curve(y_test, y_score)
+    df_roc = pd.DataFrame({'fpr': fpr, 'tpr': tpr, 'thresholds': thresholds})
+    roc_file = os.path.join('/', 'roc.csv')
+    with file_io.FileIO(roc_file, 'w') as f:
+        df_roc.to_csv(f, columns=['fpr', 'tpr', 'thresholds'], header=False, index=False)
+
+    return cm_file, roc_file, vocab
+
+
+def tf_training(args, X_train, y_train, X_test, y_test):
+    model = tf.keras.Sequential([
+        tf.keras.layers.Dense(10, activation='relu', input_shape=(4,)),  # input shape required
+        tf.keras.layers.Dense(32, activation='relu'),
+        tf.keras.layers.Dense(64, activation='relu'),
+        tf.keras.layers.Dense(128, activation='relu'),
+        tf.keras.layers.Dense(2)
+    ])
+
+    model.compile(optimizer=args.tf_nn_optimizer,
+                  loss=args.tf_nn_loss_func,
+                  metrics=[args.tf_nn_metrics])
+
+    time_hash = datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
+    log_dir = "/logs/fit/" + time_hash
+    tensorboard_callback = tf.keras.callbacks.TensorBoard(log_dir=log_dir)
+
+    model.fit(x=X_train,
+              y=y_train,
+              epochs=args.tf_nn_epoch,
+              validation_data=(X_test, y_test),
+              callbacks=[tensorboard_callback])
+
+    return model, log_dir, time_hash
+
+
+def upload_s3(args, cm_file, roc_file ,log_dir, time_hash):
+    AWS_REGION = args.aws_region
+    s3_client = boto3.client('s3', region_name=AWS_REGION)
+
+    try:
+        # upload cm file to S3
+        cm_file_name = cm_file
+        # bucket = 'eks-kubeflow-workshop-data'
+        bucket = args.s3_bucket
+        cm_object_name = 'iris-example/confusion_matrix.csv'
+        s3_cm_file = 's3://' + bucket + '/' + cm_object_name
+        s3_client.upload_file(cm_file_name, bucket, cm_object_name)
+
+        roc_object_name = 'iris-example/roc.csv'
+        s3_roc_file = 's3://' + bucket + '/' + roc_object_name
```

Comment:
Can you use file path join rather than string concatenation?

Reply:
Sure, great point. Let me fix it